client/extensions/woocommerce/components/sparkline/index.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * External dependencies
+ */
+import React, { PropTypes, Component } from 'react';
+import classNames from 'classnames';
+import { extent as d3Extent } from 'd3-array';
+import { scaleLinear as d3ScaleLinear } from 'd3-scale';
+import { select as d3Select } from 'd3-selection';
+import { line as d3Line } from 'd3-shape';
+
+/**
+ * Internal dependencies
+ */
+
+export default class Sparkline extends Component {
+	constructor( props ) {
+		super( props );
+		this.handleResize = this.handleResize.bind( this );
+		this.redrawChart = this.redrawChart.bind( this );
+		this.setNodeRef = this.setNodeRef.bind( this );
+	}
+
+	static propTypes = {
+		aspectRatio: PropTypes.number,
+		className: PropTypes.string,
+		data: PropTypes.array.isRequired,
+		highlightIndex: PropTypes.number,
+		highlightRadius: PropTypes.number,
+		margin: PropTypes.object,
+	};
+
+	static defaultProps = {
+		aspectRatio: 4.5,
+		highlightRadius: 3.5,
+		margin: {
+			top: 4,
+			right: 4,
+			bottom: 4,
+			left: 4,
+		},
+	};
+
+	state = {
+		width: 0,
+		height: 0,
+		xScale: {},
+		yScale: {},
+	};
+
+	componentDidMount() {
+		window.addEventListener( 'resize', this.handleResize );
+		this.handleResize();
+	}
+	componentDidUpdate() {
+		this.redrawChart();
+	}
+	// Remove listener
+	componentWillUnmount() {
+		window.removeEventListener( 'resize', this.handleResize );
+	}
+
+	handleResize() {
+		const { aspectRatio, data, margin } = this.props;
+		const newWidth = this.node.offsetWidth;
+		const newHeight = newWidth / aspectRatio;
+		this.setState( {
+			width: newWidth,
+			height: newHeight,
+			xScale: d3ScaleLinear()
+				.domain( d3Extent( data, ( d, i ) => i ) )
+				.range( [ margin.left, newWidth - margin.right ] ),
+			yScale: d3ScaleLinear()
+				.domain( d3Extent( data, ( d ) => d ) )
+				.range( [ newHeight - margin.bottom, margin.top ] ),
+		}, this.redrawChart );
+	}
+
+	redrawChart() {
+		const { height, width } = this.state;
+		d3Select( this.node ).selectAll( 'svg' ).remove();
+		const newNode = d3Select( this.node )
+			.append( 'svg' )
+			.attr( 'class', 'sparkline__viewbox' )
+			.attr( 'viewBox', `0 0 ${ width } ${ height }` )
+			.attr( 'preserveAspectRatio', 'xMidYMid meet' )
+			.append( 'g' );
+		this.drawSparkline( newNode );
+		if ( typeof this.props.highlightIndex !== 'undefined' ) {
+			this.drawHighlight( newNode );
+		}
+	}
+
+	drawSparkline( context ) {
```

Comment:
alternatively you could make it a top level function and pass it the xScale and yScale off state.